mdos/src/tools/vmdis.1
.\"
.\" Mach Operating System
.\" Copyright (c) 1991,1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\" 
.\" CARNEGIE MELLON ALLOWS FREE USE OF THIS SOFTWARE IN ITS 
.\" CONDITION.  CARNEGIE MELLON DISCLAIMS ANY LIABILITY OF ANY KIND FOR
.\" ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.
.\" 
.\" Carnegie Mellon requests users of this software to return to
.\" 
.\"  Software Distribution Coordinator  or  Software.Distribution@CS.CMU.EDU
.\"  School of Computer Science
.\"  Carnegie Mellon University
.\"  Pittsburgh PA 15213-3890
.\" 
.\" any improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" HISTORY
.\" $Log:	vmdis.1,v $
.\" Revision 2.3  92/04/29  16:31:42  grm
.\" 	Change the title from vmread to vmdis.
.\" 	[92/04/21            grm]
.\" 
.\" Revision 2.2  91/12/05  16:45:19  grm
.\" 	Created.
.\" 	[91/12/05  14:35:13  grm]
.\" 
.TH VMDIS 1 8/29/91
.CM 4
.SH NAME
vmdis \- read a task's address space and disassemble it as 8086 code.
.SH SYNOPSIS
\fBvmdis\fP [\fB-p] pid [addr] [size]
.SH DESCRIPTION
\fIvmdis(1)\fR reads and disassembles a task's address space as 8086
assembly instructions.
The \fIpid\fP argument, a unix server pid, specifies the task.
The \fIaddr\fP argument is a hexadecimal number
specifying the starting address.  This number may be an
absolute address, or an 8086 style address using a colon
to separate the segent and the offset.
The \fIsize\fP argument is a hexadecimal number
specifying the number of bytes of memory to disassemble.
.PP
You must either run as root or be the task's owner to gain access to
the target task.
.PP
\fIvmdis\fP interprets the following options:
.TP 8
.B \-p\fR
Memory will be interpreted as 80386 instructions.
.SH "SEE ALSO"
ps(1), vmread(1)


gnu-osfmig/NEWS
-*- mode: outline -*- Time-stamp: "2003-08-25 20:13:52 jemarch"


* New in gnu-osfmig 0.01

** Exist as a separate package, out of the osfmk sources and building framework.

** All ODE stuff dissapears

** [autoconf]igurable


mdos/src/tools/vmdis.c
/*
 * Copyright (c) 1991 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * CARNEGIE MELLON ALLOWS FREE USE OF THIS SOFTWARE IN ITS "AS IS"
 * CONDITION.  CARNEGIE MELLON DISCLAIMS ANY LIABILITY OF ANY KIND FOR
 * ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.
 *
 * Carnegie Mellon requests users of this software to return to
 *
 *  Software Distribution Coordinator  or  Software.Distribution@CS.CMU.EDU
 *  School of Computer Science
 *  Carnegie Mellon University
 *  Pittsburgh PA 15213-3890
 *
 * any improvements or extensions that they make and grant Carnegie Mellon
 * the rights to redistribute these changes.
 */
/*
 *
 * Dissassemble a block of instructions at a virtual address.
 *
 * HISTORY:
 * $Log:	vmdis.c,v $
 * Revision 2.2  91/12/05  16:45:24  grm
 * 	Changed copyrights.
 * 
 * 	Initial version grm (Gerald Malan) 90/03/09
 * 	[91/12/04            grm]
 * 
 *
 */

#include <stdio.h>
#include <sys/types.h>
#include <mach.h>

#define EIP_MODE	1
#define CSIP_MODE	2
#define PROT_MODE	3

#define MACH_CALL(x,y)	{int foo;if((foo=(x))!=KERN_SUCCESS){\
                         mach_error(y,foo);exit(1);}}

void dis_block (mode, address, cs, ip, size)
	int	mode;
	unsigned char * address;
	int	cs;
	int	ip;
	int	size;
{
	char outbuf[1024];
	unsigned char * eip;
	int numbytes;

	eip = address;
	while (eip < address + size) {
		switch (mode) {
		    case CSIP_MODE:
			fprintf(stdout, "dis: 0x%4.4x:%4.4x     ",cs,ip);
			numbytes = i386dis (mode, cs, ip, eip, outbuf);
			fprintf(stdout, "%s\n", outbuf);
			eip += numbytes;
			ip += numbytes;
			if (ip > 0xffff) {
				ip &= 0xffff;
				cs += 0x1000;
			}
			break;
		    case EIP_MODE:
		    case PROT_MODE:
		    default:
			fprintf(stdout, "dis: 0x%8.8x     ",ip);
			numbytes = i386dis (mode, cs, ip, eip, outbuf);
			fprintf(stdout, "%s\n", outbuf);
			eip += numbytes;
			ip += numbytes;
			break;
			
		}
	}
}

main(argc, argv)
	int argc;
	char **argv;
{
	int mode = 0;
	int offset;
	int pid;
	vm_size_t size, read_amount;
	kern_return_t error;
	vm_address_t addr, address, staddr, edaddr;
	char addrstr[256];
	int cs;
	int ip;
	
	task_t task;

	argc--, argv++;
	
	if ((argc > 1) && *argv[0] == '-') {
		if (strcmp(*argv, "-p") == 0) {
			mode = PROT_MODE;
			argc--; argv++;
		}
	}

	if (argc<1) {
		printf("usage: vmdis [-p] pid [addr] [size]\n");
		exit(1);
	}
	
	pid = atoi(*argv);
	argc-- , argv++;
	if (argc<1) {
		addr = 0x00;
	} else {
		char * pos;

		strcpy(addrstr, *argv);

		pos = (char *)strchr(addrstr, ':');
		if (pos) {
			*(pos++) = '\0';
			sscanf(addrstr, "%x", &cs);
			sscanf(pos, "%x", &ip);
			mode = CSIP_MODE;
			addr = (vm_address_t)(((u_short)cs << 4) + (u_short)ip);
		}else{
			cs = 0;
			sscanf(addrstr, "%x", &ip);
			if (mode != PROT_MODE)
				mode = EIP_MODE;
			addr = ip;
		}
	}

	argc-- , argv++;
	if (argc<1) {
		size = 0x50;
	} else {
		sscanf(*argv, "%x", &size);
	}

	argc-- , argv++;
	if (argc<1) {
		offset = 0x00;
	} else {
		sscanf(*argv, "%x", &offset);
	}

	task = task_by_pid(pid);

	staddr = trunc_page(addr);
	edaddr = round_page(addr+size);
	MACH_CALL((vm_read(task, staddr, edaddr - staddr,
			&address,&read_amount)),"vm_read");


	dis_block (mode, address+(addr%vm_page_size), cs, ip, size);

}
